crates/store/src/filesystem/mod.rs

```diff
@@ -0,0 +1,412 @@
+const META_FILE: &str = "meta.bincode";
+const CONTENT_LINK_FILE: &str = "content.bin";
+
+/// Filesystem backed storage.
+///
+/// Layout:
+///
+///```text
+///root/
+///  storename/
+///    sha256/
+///      key/
+///        (key.bincode checksum)/
+///          key.bincode - unique
+///          meta.bincode - possible duplicate
+///          content.bin - symlink to value/*.bin file
+///      value/
+///        (checksum).bin - unique
+///```
+#[derive(Clone)]
+pub struct Filesystem<K> {
+    root: Dir,
+    key_root: String,
+    value_root: String,
+    phantom: PhantomData<K>,
+}
+
+impl<K> Filesystem<K> {
+    async fn get_keys(&self) -> Vec<Result<K, Error>>
+    where
+        K: DeserializeOwned + 'static + Sync + Send + Clone,
+    {
+        let mut keys = vec![];
+
+        if let Ok(entries) = self.root.read_dir(&self.key_root).await {
+            for key_dir in entries {
+                if let Ok(Ok(key_dir)) = key_dir.map(|dir| dir.open_dir()) {
+                    if let Ok(entries) = key_dir.entries().await {
+                        for key_file in entries {
+                            let key_file = match key_file {
+                                Err(_) => continue,
+                                Ok(key_file) => key_file,
+                            };
+
+                            if utility::dir_entry_contains(&key_file, "key.bincode") {
+                                let key_file = key_file.open();
+
+                                keys.push(async {
+                                    self.read_bincode_file(&mut key_file.map_err(Error::Io)?)
+                                        .await
+                                });
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        join_all(keys.into_iter()).await
+    }
+
+    async fn get_key_dir(&self, expected_identity: &str) -> Result<PathBuf, Error>
+    where
+        K: Serialize + Sync + Send + Eq + Hash,
+    {
+        let key_dir: PathBuf = format!("{}/{}", self.key_root, expected_identity).into();
+
+        if !self.root.exists(&key_dir).await {
+            return Err(Error::NotFound);
+        }
+
+        let mut key_file = key_dir.clone();
+        key_file.push(KEY_FILE);
+
+        let actual_identity = utility::sha256(&self.read_bytes(&key_file).await?);
+
+        if expected_identity != actual_identity {
+            return Err(Error::InvalidKeyFile);
+        }
+
+        Ok(key_dir)
+    }
+
+    async fn read_bytes<P: AsRef<Path>>(&self, path: P) -> Result<Vec<u8>, Error> {
+        if !self.root.exists(&path).await {
+            Err(Error::NotFound)
+        } else {
+            self.root.read(path).await.map_err(Error::Io)
+        }
+    }
+
+    async fn read_bytes_file(&self, file: &mut File) -> Result<Vec<u8>, Error> {
+        let mut buffer = vec![];
+        file.read_to_end(&mut buffer).await.map_err(Error::Io)?;
```

Comment:
What is the purpose of this whole method? If it is only necessary for wrapping the error, then I'm not sure we need it.